### DISCUSSION TITLE
Safe / canonical /recommended way to merge a set of level sets that may overlap...

### DISCUSSION THREAD

**Discussion by kdmarrett** [AcademySoftwareFoundation/openvdb#1712] [Q&A]
from the C++ API?

currently I'm doing:
```
grid->tree().merge(..., vb::MERGE_ACTIVE_STATES_AND_NODES)
```
followed by a levelSetRebuild

**Comment by kdmarrett** [comment]
seems like the suggested method would be to repeatedly use csgUnion to accumulate an aggregate grid

**Comment by danrbailey** [comment]
Hi @kdmarrett, the method you shared does work, however it's not the fastest especially as the number of level sets that you are merging increases because it's always merging pairs of levelsets.

We recently replaced the backend to `tools::csgUnion` with a new implementation that relies on merging multiple grids in one operation. You can also use this directly to merge multiple grids rather than going through csgUnion. The best example I think I can point you at is the unit tests where you can see how these methods work:

https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/openvdb/unittest/TestMerge.cc#L1136

That particular example steals from all the input level sets, but there are also examples in this set of unit tests for deep-copying const level sets that you can look at. This approach should be much faster. Hope that helps, let me know if you have more questions.

**Comment by kdmarrett** [comment]
I see, clever. Thanks for the pointer, and also for posting your SIGGRAPH slides, very helpful.
